### PR TITLE
Numeric Slider Support

### PR DESCRIPTION
**Pull request by BryonLewis**
- models.py updated the attributes for optional `editor` object which can include future visualization options for attributes
- Added in the `color` option for the attribute timeline/filtering tools
- Added selection between input box/slider for numeric attributes

https://user-images.githubusercontent.com/61746913/184219479-56bc1e5d-d2ad-4821-9b73-8820a0cf98d3.mp4

### CODE REVIEW

File: client/dive-common/components/AttributeEditor.vue [hunk 1]
```diff
@@ -128,14 +162,71 @@ export default defineComponent({
...
               { text: 'Boolean', value: 'boolean' },
               { text: 'Number', value: 'number' },
               { text: 'Text', value: 'text' }
             ]"
             label="Datatype"
+            @change="typeChange"
           />
+          <div v-if="datatype=== 'number'">
+            <v-radio-group
+              :value="editor.type"
```
Comment (marySalvi) [inline client/dive-common/components/AttributeEditor.vue:176]: I think this line is causing an error for existing attributes. When you click the edit attribute cog the dialog fails to load and console displays this:
image

Comment (BryonLewis) [reply client/dive-common/components/AttributeEditor.vue:176]: Yup, editor probably needs to be checked to see if it exists, editor can be undefined.

File: client/dive-common/components/AttributeEditor.vue [hunk 2]
```diff
@@ -128,14 +162,71 @@ export default defineComponent({
...
+                hint="Lower limit for slider"
+                persistent-hint
+              />
+              <v-text-field
+                v-model.number="editor.range[1]"
+                dense
+                outlined
+                :step="editor.range[1]> 1 ? 1 : 0.01"
+                type="number"
+                label="Upper"
```
Comment (marySalvi) [inline client/dive-common/components/AttributeEditor.vue:210]: I would change label to 'Max'

File: client/dive-common/components/AttributeEditor.vue [hunk 3]
```diff
@@ -128,14 +162,71 @@ export default defineComponent({
...
+          </div>
+          <div v-if="datatype === 'number' && editor && editor.type === 'slider'">
+            <v-row class="pt-2">
+              <v-text-field
+                v-model.number="editor.range[0]"
+                dense
+                outlined
+                :step="editor.range[0]> 1 ? 1 : 0.01"
+                type="number"
+                label="Lower"
```
Comment (marySalvi) [inline client/dive-common/components/AttributeEditor.vue:199]: I would change label to 'Min'

File: client/dive-common/components/AttributeEditor.vue [hunk 4]
```diff
@@ -128,14 +162,71 @@ export default defineComponent({
...
+              />
+            </v-row>
+            <v-row class="pt-2">
+              <v-text-field
+                v-model.number="editor.steps"
+                dense
+                outlined
+                :step="editor.steps> 1 ? 1 : 0.01"
+                type="number"
+                label="Slider Resolution"
```
Comment (marySalvi) [inline client/dive-common/components/AttributeEditor.vue:223]: In would change label to something like 'Slider Step Interval' and the hint to something like 'Slider will increase/decrease by this value'

File: client/dive-common/components/AttributeInput.vue [hunk 5]
```diff
@@ -143,6 +154,23 @@ export default defineComponent({
...
+    <div
+      v-else-if="datatype === 'number' && (typeSettings && typeSettings.type ==='slider')"
+    >
+      <div class="slider-label">
+        {{ value }}
+      </div>
+      <v-slider
+        :value="value"
+        :step="typeSettings.steps ? typeSettings.steps
+          : (typeSettings.range[1] - typeSettings.range[0])/2.0"
```
Comment (marySalvi) [inline client/dive-common/components/AttributeInput.vue:166]: It's possible for both `typeSettings.steps` and `typeSettings.range` to be undefined and the slider breaks. Probably best to make one or the other mandatory when creating the slider.
Screenshot from 2022-08-16 13-30-58

 
Screenshot from 2022-08-16 13-30-37

File: client/dive-common/components/AttributeInput.vue [hunk 6]
```diff
@@ -143,6 +154,23 @@ export default defineComponent({
...
+        :min="typeSettings.range[0]"
+        :max="typeSettings.range[1]"
```
Comment (marySalvi) [inline client/dive-common/components/AttributeInput.vue:168]: We either need to make the range required or have a default value

**Review by marySalvi — CHANGES_REQUESTED** [review]
When changing a combobox with an existing value to a slider, the value is lost. Is that behavior we want or would we want the slider positioned at that previous value?

Screenshot from 2022-08-16 13-49-30

**Comment by BryonLewis** [comment]
> When changing a combobox with an existing value to a slider, the value is lost. Is that behavior we want or would we want the slider positioned at that previous value?
> 
> Screenshot from 2022-08-16 13-49-30

I think that's okay because I can't guarantee that the new range contains the old value.  We auto set a range when they swap so that is why the value changes.  Ideally this should be set up in the configuration file once and then values are undefined/0 until the user sets them on a specific track/detection attribute.

**Review by marySalvi — APPROVED** [review]
Looks Great!

**Pull request merged**